spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
         self.total_size += len(chunk)
-        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+        if self.total_size > self._maximum_allowed_size:
             raise ArchiveTooLargeError()
 
-    def read(self, size=-1) -> bytes:
-        content = self._trailing_data
-        self._trailing_data = b''
+    def _num_bytes_left_in_buffer(self) -> int:
+        return len(self._buffer) - self._buffer_position + 1
 
-        if not size or size == -1:
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-        else:
-            # iterate chunks until we've read at least `size` amount of data
+    def read(self, size=-1) -> bytes | bytearray | memoryview:
+        if self._consumed:
+            return b''
+
+        elif not size or size == -1:
             for chunk in self._chunks:
                 self._add_chunk_to_size(chunk)
-                content += chunk
-                if len(content) > size:
-                    break
+                self._buffer += chunk
 
-            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
-            if len(content) > size:
-                self._trailing_data = content[size:]
-                content = content[:size]
+            self._consumed = True
+            return self._buffer
 
-        return content
+        else:
+            # Check if the data left in buffer if enough to satisfy the full read size...
+            if self._num_bytes_left_in_buffer() < size:
+                # If not, chop off everything that we've read out of the buffer, and start re-filling it. Reading
+                #  from our memoryview is actually slower here, since we end up having to convert that to bytes
+                #  anyway for our concatenation down below
+                self._buffer = self._buffer[self._buffer_position:]
+                self._buffer_position = 0
+
+                # iterate chunks until we've read at least `size` amount of data
+                for chunk in self._chunks:
+                    self._add_chunk_to_size(chunk)
+                    self._buffer += chunk
+                    if len(self._buffer) >= size:
+                        break
+
+                # Freeze this buffer for subsequent reads from it. We use a memoryview in order to avoid making copies
+                #  of the data for every single read issued.
+                self._buffer_memview = memoryview(self._buffer).cast("B").toreadonly()
+
+            # If we have enough data in the buffer to satisfy this read, return that from the buffer and adjust
+            #  our read position accordingly
+            if self._num_bytes_left_in_buffer() >= size:
+                start_pos = self._buffer_position
+                self._buffer_position += size
+                return self._buffer_memview[start_pos:self._buffer_position]
```

Comment:
Related to the above comment - `self._buffer[...]` would return a _copy_ of that data, whereas `self._buffer_memview[...]` returns the data without copying it.